clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -709,12 +711,25 @@ private boolean appendsInProgress() {
     }
 
     /**
-     * Mark all partitions as ready to send and block until the send is complete
+     * Mark all partitions as ready to send and block until the send is complete or time expires
      */
-    public void awaitFlushCompletion() throws InterruptedException {
+    public void awaitFlushCompletion(long timeoutMs) throws InterruptedException {
         try {
-            for (ProducerBatch batch : this.incomplete.copyAll())
-                batch.produceFuture.await();
+            Long expireMs = System.currentTimeMillis() + timeoutMs;
+            int numBatchesFlushed = 0;
+            for (ProducerBatch batch : this.incomplete.copyAll()) {
+                Long currentMs = System.currentTimeMillis();
+                if (currentMs > expireMs) {
+                    throw new TimeoutException(String.format("Failed to flush accumulated records within %d milliseconds,"
+                        + " successfully completed %d batches.", timeoutMs, numBatchesFlushed));
+                }
```

Comment:
Is this strictly necessary? i.e., you are bounding the subsequent `await` to not pass the deadline right?